sdks/go/pkg/beam/core/runtime/exec/fn_test.go

```diff
@@ -400,3 +400,210 @@ func BenchmarkInvokeFnCallExtra(b *testing.B) {
+
+	var a int
+	var ai interface{} = a
+	aV := reflect.ValueOf(a)
+	rvSlice := []reflect.Value{aV}
+	grvSlice := []reflect.Value{gV, aV}
+	efaceSlice := []interface{}{a}
+	gEfaceSlice := []interface{}{g, a}
+
+	tests := []struct {
+		name string
+		fn   func()
+	}{
+		{"DirectMethod", func() { a = g.WhatsA(a) }}, // Baseline as low as we can go.
+		{"DirectFunc", func() { a = WhatsB(a) }},     // For comparison purposes
+
+		{"IndirectFunc", func() { a = indirectFunc(a) }},         // For comparison purposes
+		{"IndirectImplicit", func() { a = nrFi(a) }},             // Measures the indirection through reflect.Value cost.
+		{"TypeAssertedImplicit", func() { ai = nrFi(ai.(int)) }}, // Measures the type assertion cost over the above.
+
+		{"ReflectCallImplicit", func() { a = nrF.Call([]reflect.Value{reflect.ValueOf(a)})[0].Interface().(int) }},
+		{"ReflectCallImplicit-NoWrap", func() { a = nrF.Call([]reflect.Value{aV})[0].Interface().(int) }},
+		{"ReflectCallImplicit-NoReallocSlice", func() { a = nrF.Call(rvSlice)[0].Interface().(int) }},
+
+		{"ReflectXCallImplicit", func() { a = rxnrF.Call([]interface{}{a})[0].(int) }},
+		{"ReflectXCallImplicit-NoReallocSlice", func() { a = rxnrF.Call(efaceSlice)[0].(int) }},
+		{"ReflectXCall1x1Implicit", func() { a = rx0x1nrF.Call1x1(a).(int) }}, // Measures the default shimfunc overhead.
+
+		{"ShimedCallImplicit", func() { a = shimnrF.Call([]interface{}{a})[0].(int) }},          // What we're currently using for invoking methods
+		{"ShimedCallImplicit-NoReallocSlice", func() { a = shimnrF.Call(efaceSlice)[0].(int) }}, // Closer to what we're using now.
+		{"ShimedCall1x1Implicit", func() { a = shim0x1nrF.Call1x1(a).(int) }},
+
+		{"IndirectExplicit", func() { a = wrFi(g, a) }},                     // Measures the indirection through reflect.Value cost.
+		{"TypeAssertedExplicit", func() { ai = wrFi(gi.(*Foo), ai.(int)) }}, // Measures the type assertion cost over the above.
+
+		{"ReflectCallExplicit", func() { a = wrF.Call([]reflect.Value{reflect.ValueOf(g), reflect.ValueOf(a)})[0].Interface().(int) }},
+		{"ReflectCallExplicit-NoWrap", func() { a = wrF.Call([]reflect.Value{gV, aV})[0].Interface().(int) }},
+		{"ReflectCallExplicit-NoReallocSlice", func() { a = wrF.Call(grvSlice)[0].Interface().(int) }},
+
+		{"ReflectXCallExplicit", func() { a = rxF.Call([]interface{}{g, a})[0].(int) }},
+		{"ReflectXCallExplicit-NoReallocSlice", func() { a = rxF.Call(gEfaceSlice)[0].(int) }},
+		{"ReflectXCall2x1Explicit", func() { a = rx1x1F.Call2x1(g, a).(int) }},
+
+		{"ShimedCallExplicit", func() { a = shimF.Call([]interface{}{g, a})[0].(int) }},
+		{"ShimedCallExplicit-NoReallocSlice", func() { a = shimF.Call(gEfaceSlice)[0].(int) }},
+		{"ShimedCall2x1Explicit", func() { a = shim1x1F.Call2x1(g, a).(int) }},
+	}
+	for _, test := range tests {
+		b.Run(test.name, func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				test.fn()
+			}
+		})
+	}
+	b.Log(a)
+}
+
+/*
```

Comment:
Is there value in keeping this information here in the comments? It is unlikely that it would be useful to others, other than you because each desktop config will be unique. (Consider moving to wiki, or running an standard GCE VM and adding data based on that.)

Reply:
This keeps a record of *why* a decision was made, and allows us to see what changed over time, and to keep us honest if for whatever reason reflection becomes sufficiently fast to avoid dealing with this altogether. Putting it in the wiki is a good idea, but I'd prefer to do that once I set my sights making the Wiki useful for the Go SDK. There's too much pre work I'd need to do before I could just move these numbers there.

The precise measurements are going to change, but their relative ordering and magnitudes will probably not, short of differences between Go versions. This actually happened with Go 1.10 -> Go 1.11 where method invocation with the Implicit Receiver improved by ~50ns/op since they started doing better caching of the implicit receivers. If a complete re-write to use Explicit Receivers takes long enough, it could be that Go removes the distinction performance wise, which would let us keep the simpler code.